### ISSUE TITLE
Slow scrolling in clojure-mode buffers

### ISSUE THREAD

**Issue by kommen** [syl20bnr/spacemacs#10939]
#### Description :octocat:
Scrolling down clojure buffers in my project is very laggy and puts the emacs CPU usage at 100%. 

#### Reproduction guide :beetle:
- Start Emacs
- open a large Clojure or .edn file 
- position point at the last visible line in the buffer
- move point down so the buffer has to scroll

*Observed behaviour:* :eyes: :broken_heart:
Scrolling is very laggy.

*Expected behaviour:* :heart: :smile:
Smooth scrolling.

#### System Info :computer:
- OS: darwin
- Emacs: 26.1
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. f1f8db002)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(elixir javascript html org clojure helm auto-completion emacs-lisp git osx markdown version-control yaml github)
```
- System configuration features: JPEG RSVG IMAGEMAGICK NOTIFY ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS NS MODULES THREADS LCMS2

#### CPU profile

The profiler hints that spacemacs/title-prepare could some how be the culprit:

```
- redisplay_internal (C function)                                 149  79%
 - eval                                                           149  79%
  - spacemacs/title-prepare                                       111  59%
   - let*                                                         111  59%
    - format-spec-make                                            111  59%
       abbreviate-file-name                                       111  59%
  - spaceline-ml-main                                              37  19%
   - powerline-width                                               17   9%
    - powerline-width                                              17   9%
     + powerline-width                                             16   8%
   + mapcar                                                         6   3%
     powerline-raw                                                  3   1%
   + powerline-hud                                                  2   1%
     file-remote-p                                                  1   0%
   + s-trim                                                         1   0%
   + run-hooks                                                      1   0%
     powerline-render                                               1   0%
   + powerline-wave-left                                            1   0%
+ timer-event-handler                                              23  12%
+ command-execute                                                  15   8%
+ ...                                                               0   0%
```

**Comment by kommen** [comment]
I have `dotspacemacs-frame-title-format` configured to be `%b`.

**Comment by hackeryarn** [comment]
Are you running `emacs-plus` or `emacs --with-cocoa`? I noticed a big difference when I switched to `emacs-plus` on my Mac.

**Comment by kommen** [comment]
I'm using `emacs-plus` installed via brew.

**Comment by Miciah** [comment]
I use Emacs on X.Org on GNU/Linux, and I've seen similar performance issues from `spacemacs/title-prepare`, so I set `dotspacemacs-frame-title-format` and `dotspacemacs-icon-title-format` both to `nil`.

Spacemacs sets `frame-title-format` and `icon-title-format` to the respective `dotspacemacs-` variables on startup, so you can either set the `dotspacemacs-` variables to `nil` and restart to test the performance impact or set the non-`dotspacemacs-` variables to `nil` to test without restarting.

**Comment by Compro-Prasad** [comment]
@kommen So, it solves your issue? Right?

**Comment by Miciah** [comment]
It would be nice not to need to disable title bars and modeline segments to get acceptable responsiveness.  Ideally, the performance problems in Powerline/Spaceline would be fixed.  Next best would be to have defaults that avoided these performance problems.

**Comment by Compro-Prasad** [comment]
There are many issues related to spaceline and it has already been reported upstream. https://github.com/syl20bnr/spacemacs/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+spaceline+freeze shows some of the results.

**Comment by kommen** [comment]
@Compro-Prasad no, unfortunately I still see the laggy scrolling. Not as bad as before, but still jarring.  I tried disabling the modeline completely, but that didn't improve the lag further. (maybe semi-relevant: this is on a 2017 15" MacBook Pro with 2,8 GHz i7). Just ran a new profile, and it looks much like the one I posted in https://github.com/syl20bnr/spacemacs/issues/10939#issuecomment-399686433, which blames `auto-revert-buffers`.

**Comment by kommen** [comment]
Also what makes me curios is why I have this performance issue mostly in clojure buffers. E.g in we also have a elisp file in that project, where scrolling is noticeable smoother, even with the tweaks above reverted.

**Comment by Compro-Prasad** [comment]
@kommen Best you can do is strat from scratch and build up your own Emacs config taking small bits from spacemacs. This will help you to separate out the package or config option.

**Comment by kommen** [comment]
I took on this issue and it seems I found a solution. This has nothing to do with clojure mode directly. I'm on a MacBook Pro (15-inch, 2017), sometimes with a 5K display connected. Emacs is just overwhelmed it seems pushing all those pixels, running on 100% cpu when scrolling, while the emacs profile doesn't really report anything suspicious.

https://bitbucket.org/mituharu/emacs-mac/src to the rescue. It comes with support for [Metal](https://en.wikipedia.org/wiki/Metal_%28API%29) to push all those pixels much more effectively. I see **huge** performance improvements using that instead of `emacs-plus`. One small caveat: the pre build binary from brew cask ships without Metal support, one has to build it from source, see https://github.com/railwaycat/homebrew-emacsmacport/issues/155

**Comment by kommen** [comment]
Also: some performance improvements also came from this before: https://github.com/syl20bnr/spacemacs/pull/11418

**Comment by Compro-Prasad** [comment]
@kommen  Is this mentioned somewhere in the readme or faq? Should be there I suppose. It is the case I have seen for many Mac users. I would suggest you write it somewhere directly and easily accessible in one read.
